validator/src/journal/block_manager.rs

```diff
@@ -47,6 +48,35 @@ impl From<BlockStoreError> for BlockManagerError {
     }
 }
 
+/// An external block reference that is owned by and can be passed between validator components.
+/// Implements the `Drop` trait to allow for automatically decrementing the external reference
+/// count for the block ID in the block manager.
+pub struct BlockRef {
+    block_id: String,
+    block_manager: BlockManager,
+}
+
+impl BlockRef {
+    fn new(block_id: String, block_manager: BlockManager) -> Self {
+        BlockRef {
+            block_id,
+            block_manager,
+        }
+    }
+
+    pub fn block_id(&self) -> String {
+        self.block_id.clone()
+    }
+}
+
+impl Drop for BlockRef {
+    fn drop(&mut self) {
+        self.block_manager
+            .unref_block(self.block_id.as_str())
+            .expect("Failed to unref block");
```

Comment:
A panic in drop() is considered bad, and may cause the process to abort. This can apparently happen if a panic happens while unwinding due to another panic.

Reply:
Agreed. I'm not sure I even agree with the fact that unref returns a result.

 Ultimately, unref_block should not be called in any way outside of this drop call in the long run, which would mean it should be managed via static garbage collection.

Reply:
Updated